### PR TITLE
Add Kimi K2 native lite support with FSDP2 validation

### PR DESCRIPTION
**Pull request by ISEEKYAN**
## Summary
- Add native Kimi K2 MLite model/protocol/checkpoint support.
- Extract Kimi's MLA into the shared primitive entrypoint `megatron.lite.primitive.attention.MultiLatentAttention`; Kimi now imports the shared primitive so GLM5/DS4 can reuse the canonical MLA base.
- Wire MTP, PP/VPP layout checks, HF weight load/export/save, FSDP2 and distopt paths for Kimi K2.

## Validation
Final head: `eea9ecb71cea3698e437c48ce72f8cc262e707fb`, rebased on `main` `281d54391f3be65637f727e6b21420f1e1109d07`.

Local checks on final head:
- `python -m black --line-length 100 --check experimental/lite/megatron/lite/model/kimi_k2 experimental/lite/megatron/lite/primitive/attention experimental/lite/tests/unit/model/test_kimi_k2_lite_static.py experimental/lite/tests/unit/model/test_kimi_k2_lite_cp_smoke.py` -> clean.
- `python -m py_compile experimental/lite/megatron/lite/primitive/attention/__init__.py experimental/lite/megatron/lite/primitive/attention/mla.py experimental/lite/megatron/lite/model/kimi_k2/lite/model.py experimental/lite/tests/unit/model/test_kimi_k2_lite_static.py experimental/lite/tests/unit/model/test_kimi_k2_lite_cp_smoke.py` -> clean.
- `python -m pytest -q experimental/lite/tests/unit/model/test_kimi_k2_lite_static.py experimental/lite/tests/unit/model/test_kimi_k2_lite_cp_smoke.py` -> `13 passed, 5 skipped`.
- `python -m pytest -q experimental/lite/tests/unit/primitive/test_dist_opt_validation.py` -> `3 passed`.
- `git diff --check is/main...HEAD` -> clean.
- PR range contains only Kimi model files, registry wiring, shared `primitive/attention/mla.py`, and Kimi tests; no `.gitignore`, `devlog/`, `primitive/modules/`, or `primitive/parallel/` changes.
- Diff adds no `megatron.core` imports; Kimi model + shared attention primitive grep has no `megatron.core` matches.

Slurm validation evidence:
- Megatron-Core fused primary parity: job `12779439`, `COMPLETED 0:0`, non-skip marker `NON_SKIP_KIMI_K2_MCORE_PARITY_PASSED`, `bitwise_all=true` for layer/final/logits/MTP diffs.
- HF secondary reference + HF weight I/O/save/reload: job `12779422`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_HF_REF_IO_PRECISION_PASSED`.
- Distopt deterministic run-to-run reproducibility: job `12779424`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_DISTOPT_R2R_DETERMINISM_PASSED`, loss/grad/MTP diffs all zero.
- Packed THD variable sequence + CP real parallel smoke: job `12797582`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_THD_CP_SMOKE_PASSED`, `world_size=2`, lengths `[5, 7, 9]`, backward included.
- 128-GPU reduced SFT smoke: job `12773371`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_SFT_COMPLETED`.

### CODE REVIEW

**Comment by ISEEKYAN** [comment]
Kimi K2 merge validation update at `dc8b173ddb2f86a46231645a9082128afd68256c`.

Scope for this validation pass: precision against independent references plus HF weight I/O. The scaled SFT run is deferred to a later owner-approved pass.

Megatron-Core primary reference:
- Slurm job: `12771248` (`kimi-mcore-ref`), `sacct` state `COMPLETED`, exit `0:0`.
- Non-skip marker: `NON_SKIP_KIMI_K2_MCORE_PARITY_PASSED`.
- Reference/candidate: Megatron-Core `GPTModel` with `MLATransformerConfig` MLA+MoE vs native MLite `KimiK2Model`.
- Weight copy: `copied=35`, `unmapped_lite=[]`; the script raises on unmapped weights or tolerance failures.
- Attention paths covered: `dense_mla_layer0`, `moe_mla_layer1`.
- Real non-zero diffs, all within tolerance:
  - layer0 hidden: max_abs `0.000244140625`, mean_abs `2.85380519926548e-06`, rms `1.6522415535291657e-05`.
  - layer1 hidden: max_abs `0.000244140625`, mean_abs `9.358394891023636e-06`, rms `3.0081742806942202e-05`.
  - final hidden: max_abs `0.015625`, mean_abs `0.0005708653479814529`, rms `0.0017802482470870018`.
  - logits: max_abs `0.00390625`, mean_abs `0.00019977320334874094`, rms `0.000408250984037295`.
- RoPE path: `yarn_interleaved_mla`; HyperConnection is not applicable for Kimi K2.

HF secondary reference and weight I/O:
- Slurm job: `12771339` (`kimi-hf-ref`), `sacct` state `COMPLETED`, exit `0:0`.
- Non-skip marker: `NON_SKIP_KIMI_K2_HF_REF_IO_PRECISION_PASSED`.
- Reference/candidate: `transformers.DeepseekV3ForCausalLM` vs native MLite `KimiK2Model`.
- `load_hf_weights` time: `0.01072273077443242s`.
- HF parity diffs, all within tolerance:
  - layer0 hidden: max_abs `0.000244140625`, mean_abs `1.3029435649514198e-05`, rms `3.62994396709837e-05`, max_rel `0.0030487803742289543`.
  - layer1 hidden: max_abs `0.0003662109375`, mean_abs `2.6967842131853104e-05`, rms `5.6462969951098785e-05`, max_rel `0.004411764908581972`.
  - logits: max_abs `0.00390625`, mean_abs `0.00038752672844566405`, rms `0.0006090219831094146`, max_rel `0.007042253389954567`.
- `export_hf_weights`: `41` tensors, `0.00011754035949707031 GiB`, `0.001331876963376999s`, `0.08825166492785079 GiB/s`.
- `save_hf_weights`: `41` tensors, `0.00012159347534179688 GiB`, `0.005163659807294607s`, `0.023547925285477563 GiB/s`.
- Router bias roundtrip: `save_router_bias_roundtrip=true`. Kimi K2 does not use a tid2eid tensor; the equivalent persisted HF roundtrip target is `model.layers.*.mlp.gate.e_score_correction_bias`.

Local focused tests:
- `PYTHONPATH=experimental/lite pytest -q experimental/lite/tests/unit/model/test_kimi_k2_lite_static.py experimental/lite/tests/unit/primitive/test_dist_opt_validation.py`
- Result: `14 passed`.

**Comment by ISEEKYAN** [comment]
Additional Kimi K2 validation update for head `8a5e4224fb4bbd7f2605385ea5c63b33085ca616`:

- MTP is now wired for native Kimi K2 lite instead of being forced off. The validation uses `num_nextn_predict_layers=1` and covers the MTP layer in both independent references.
- PP/VPP layout behavior is explicit: main transformer layers must divide the requested PP or PP*VPP layout; no padding is inserted. MTP is attached only on the head stage. Static coverage was added for the non-divisible error and head-stage MTP placement.
- Megatron-Core primary reference: Slurm job `12773235`, `COMPLETED 0:0`, non-skip marker `NON_SKIP_KIMI_K2_MCORE_PARITY_PASSED`. Deterministic settings were enabled (`NVTE_ALLOW_NONDETERMINISTIC_ALGO=0`, `CUBLAS_WORKSPACE_CONFIG=:4096:8`, `TORCHDYNAMO_DISABLE=1`, torch deterministic algorithms, and MCore deterministic mode). The run compares MCore `GPTModel`/`MLATransformerConfig` against native MLite with `copied=60`, `unmapped_lite=[]`, and paths `dense_mla_layer0`, `moe_mla_layer1`, `mtp_moe_mla_layer2`. Strict full bitwise was not reached: dense MLA layer0 is bitwise (`max_abs=0`), while the first nonzero differences start at bf16 MoE/MTP paths (`layer1 max_abs=1.2207e-4`, `final_hidden max_abs=7.8125e-3`, `logits max_abs=1.9531e-3`, `mtp0_hidden max_abs=1.5625e-2`, `mtp0_logits max_abs=3.90625e-3`). Since the mapping is complete and the dense MLA path is bitwise, the remaining non-bitwise behavior is from the independent MCore MoE/MTP kernel/order path versus MLite's native MoE/MTP path, not a self-compare or missing weights.
- HF secondary reference + weight I/O: Slurm job `12773234`, `COMPLETED 0:0`, non-skip marker `NON_SKIP_KIMI_K2_HF_REF_IO_PRECISION_PASSED`. Reference is `transformers.DeepseekV3ForCausalLM` plus an independent HF-named DeepseekV3 MTP block. Diffs are within tolerance: `layer0 max_abs=2.4414e-4`, `layer1 max_abs=3.6621e-4`, `logits max_abs=3.90625e-3`, `mtp0_hidden max_abs=1.5625e-2`, `mtp0_logits max_abs=3.90625e-3`.
- Separate `save_hf_weights` path was rerun in job `12773234`: `save_keys=71`, `save_seconds=0.0061417`, `save_gib_per_s=0.02948`. Exported weight tensors are bf16; only router `e_score_correction_bias` remains allowed fp32. Reloading the saved HF export into a fresh MLite model gives `reload_logits max_abs=0` and `reload_mtp0_logits max_abs=0`. Kimi uses router correction bias, so tid2eid roundtrip is not applicable.
- 128-GPU scaled SFT smoke: Slurm job `12773371`, 16 nodes x 8 GPUs, `COMPLETED 0:0`, non-skip completion markers present. Config: FSDP2, EP=4, TP=1, PP=1, 4 steps. Rank0 final step: `train/loss=11.776773452758789`, `grad_norm=1.0561963319778442`, `global_tokens=23144`; validation `val/loss=11.76793384552002`. Peak GPU memory from the per-node monitor markers is `5862 MiB`.

Local checks:
- `PYTHONPATH=experimental/lite python -m pytest experimental/lite/tests/unit/model/test_kimi_k2_lite_static.py -q` -> 12 passed.
- `PYTHONPATH=experimental/lite python -m pytest experimental/lite/tests/unit/primitive/test_fsdp2_unit.py experimental/lite/tests/unit/primitive/test_parallel_unit.py -q` -> 25 passed, 1 skipped.
- `git diff --check` -> clean.

**Comment by ISEEKYAN** [comment]
## MLite Kimi K2 Merge Validation Gate Checklist

Head: `7b9b81793cc2967305fa41e684cbb6e8f68e474b7` rebased on `is/main` `67c30d34669249c83d460803b181ef144e5ed621`.

### Precision Alignment
- [x] ✅ forward fused-vs-fused: MLite fused vs Megatron-Core corresponding fused path, same weights, layer/final/logits/MTP diffs bitwise zero; `unmapped_lite=[]`; fused counters prove non-eager path (`mcore_te_core_attention=3`, `lite_te_core_attention=3`, `mcore_te_grouped_mlp=2`, `lite_moe_experts=2`, `mcore_weighted_bias_swiglu=2`, `lite_weighted_bias_swiglu=2`) — Slurm job `12779439`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_MCORE_PARITY_PASSED`.
- [x] ✅ HF secondary reference: vs `transformers.DeepseekV3ForCausalLM` plus HF-named MTP block, failures `{}`; diffs within tolerance (`layer0 max_abs=2.4414e-4`, `layer1 max_abs=3.6621e-4`, `logits max_abs=3.90625e-3`, `mtp0_hidden max_abs=1.5625e-2`, `mtp0_logits max_abs=3.90625e-3`) — Slurm job `12779422`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_HF_REF_IO_PRECISION_PASSED`.
- [x] ✅ run-to-run deterministic reproducibility: distopt + deterministic two identical train steps, `loss/mtp_loss/grad_norm/dense_grad_values/mtp_grad_values` max_abs all `0` — Slurm job `12779424`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_DISTOPT_R2R_DETERMINISM_PASSED`.

### Key Features
- [x] ✅ MTP: `num_nextn_predict_layers=1`, MTP path covered against both Megatron-Core primary and HF secondary references; MCore job has `mtp0_hidden=0`, `mtp0_logits=0`; HF job covers `mtp0_hidden` and `mtp0_logits` within tolerance — jobs `12779439`, `12779422`.
- [ ] ❌ THD + CP: no 32-GPU/128k seqlen FSDP2 EP8 CP8 allgather-vs-A2A benchmark evidence in this Kimi validation pass.
- [ ] ❌ proxy smoke: no dedicated FSDP2+EP+PP+CP non-skip proxy training-step job in this Kimi validation pass.
- [x] ✅ end-to-end reduced 128-GPU SFT: 16 nodes x 8 GPUs, 4 steps, FSDP2 EP4; non-skip completion markers present; peak memory marker max `5862 MiB` — Slurm job `12773371`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_SFT_COMPLETED`.

### Weight I/O
- [x] ✅ `load_hf_weights` / `export_hf_weights` / `save_hf_weights` correctness and speed: `load_hf_seconds=0.0206656`, `export_count=71`, `export_seconds=0.0020800`, `export_gib_per_s=0.08365`, `save_keys=71`, `save_seconds=0.0062735`, `save_gib_per_s=0.02886`, router bias roundtrip true; Kimi tid2eid is N/A because it uses router `e_score_correction_bias` — job `12779422`.
- [x] ✅ separate `save_hf_weights` bf16 + reload: all weight tensors bf16 except allowed fp32 router correction bias; fresh reload `reload_logits max_abs=0`, `reload_mtp0_logits max_abs=0` — job `12779422`.

### Structure / Engineering
- [x] ✅ PP non-divisible behavior and tests: main layers must divide PP or PP*VPP; no padding; MTP attaches only on head stage. Static tests rechecked after rebase with `PYTHONPATH=experimental/lite python -m pytest -q experimental/lite/tests/unit/model/test_kimi_k2_lite_static.py experimental/lite/tests/unit/model/test_kimi_k2_lite_cp_smoke.py experimental/lite/tests/unit/primitive/test_dist_opt_validation.py` -> `16 passed, 4 skipped`.
- [x] ✅ PR hygiene: rebased on the landed shared-primitive cleanup; `.gitignore` is not in the final PR range; no `devlog/` files are tracked; `python -m black --check --line-length 100 ...` clean; `git diff --check is/main` clean.
- [x] ✅ non-kernel primitive mcore-import cleanup inherited: final diff adds no `megatron.core` imports outside the expected distopt/distckpt integration areas (`git diff is/main...HEAD -U0 -- 'experimental/lite/**/*.py' | rg '^\+.*(from megatron\.core|import megatron\.core|megatron\.core)'` returns no matches). Kimi `model.py` now uses the lite vendored `megatron.lite.primitive.kernels.swiglu` path.
- [x] ✅ distopt naming: Kimi model path uses `distopt`; remaining changed distopt wrapper/test references are the intended Megatron-Core optimizer integration points.

**Comment by ISEEKYAN** [comment]
## MLite Kimi K2 Merge Validation Gate Checklist

Head: `eea9ecb71cea3698e437c48ce72f8cc262e707fb`, rebased on post-PR#30 `main` `281d54391f3be65637f727e6b21420f1e1109d07`.

### Precision Alignment
- [x] ✅ forward fused-vs-fused: MLite fused vs Megatron-Core fused path, same weights, layer/final/logits/MTP diffs bitwise zero; `unmapped_lite=[]` — Slurm job `12779439`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_MCORE_PARITY_PASSED`.
- [x] ✅ HF secondary reference: vs `transformers.DeepseekV3ForCausalLM` plus HF-named MTP block; failures `{}`; diffs within tolerance — Slurm job `12779422`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_HF_REF_IO_PRECISION_PASSED`.
- [x] ✅ run-to-run deterministic reproducibility: distopt + deterministic two identical train steps, `loss/mtp_loss/grad_norm/dense_grad_values/mtp_grad_values` max_abs all `0` — Slurm job `12779424`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_DISTOPT_R2R_DETERMINISM_PASSED`.

### Key Features
- [x] ✅ MTP: `num_nextn_predict_layers=1`; MTP path covered against both Megatron-Core primary and HF secondary references — jobs `12779439`, `12779422`.
- [x] ✅ packed THD + CP real parallel smoke: variable sequence lengths `[5, 7, 9]`, CP world size `2`, forward + backward through Kimi MLA-MoE with packed THD inputs; non-skip marker present — Slurm job `12797582`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_THD_CP_SMOKE_PASSED`.
- [ ] ❌ proxy smoke: no dedicated FSDP2+EP+PP+CP non-skip proxy training-step job in this Kimi validation pass.
- [x] ✅ end-to-end reduced 128-GPU SFT: 16 nodes x 8 GPUs, 4 steps, FSDP2 EP4; non-skip completion markers present; peak memory marker max `5862 MiB` — Slurm job `12773371`, `COMPLETED 0:0`, marker `NON_SKIP_KIMI_K2_SFT_COMPLETED`.

### Weight I/O
- [x] ✅ `load_hf_weights` / `export_hf_weights` / `save_hf_weights` correctness and speed: router bias roundtrip true; Kimi tid2eid is N/A because it uses router `e_score_correction_bias` — job `12779422`.
- [x] ✅ separate `save_hf_weights` bf16 + reload: all weight tensors bf16 except allowed fp32 router correction bias; fresh reload `reload_logits max_abs=0`, `reload_mtp0_logits max_abs=0` — job `12779422`.

### Structure / Engineering
- [x] ✅ post-PR#30 rebase: final PR base is `281d54391`; PR range is one English commit.
- [x] ✅ shared MLA primitive: Kimi MLA now lives at `megatron.lite.primitive.attention.MultiLatentAttention`; `model/kimi_k2/lite/model.py` imports the shared primitive, and model-local `mla.py` is gone so GLM5/DS4 can reuse the same canonical MLA base.
- [x] ✅ PR hygiene: no `.gitignore`, no `devlog/`, no `primitive/modules/`, no `primitive/parallel/`, and no unrelated formatting churn in the PR range.
- [x] ✅ 0 mcore import: final diff adds no `megatron.core` imports; Kimi model + shared attention primitive grep has no `megatron.core` matches outside existing optimizer/checkpoint integration areas.
- [x] ✅ local checks on final head: black@100 check clean; py_compile clean; Kimi static+CP smoke local run `13 passed, 5 skipped`; distopt validation `3 passed`; `git diff --check is/main...HEAD` clean.
- [x] ✅ PP non-divisible behavior and tests: main transformer layers must divide PP or PP*VPP; no padding path is introduced; MTP attaches only on the head stage.
- [x] ✅ distopt naming: Kimi protocol/model path uses `distopt`; no old `mc` / `mc_full` optimizer naming in the Kimi diff.
